scheme.py

```diff
@@ -1,62 +1,187 @@
+from __future__ import annotations
+
 from dataclasses import dataclass
-from typing import List, Optional, Tuple, Iterable
+from typing import (
+    List, Optional, Tuple, Iterable, Sequence, Union, cast, TYPE_CHECKING)
+
+if TYPE_CHECKING:
+    from environment import Environment
 
 
-@dataclass
 class SExp:
     """An s-expression base class"""
-    pass
+    def __init__(self):
+        self._environment: Optional[Environment] = None
+
+    @property
+    def environment(self) -> Environment:
+        if self._environment is None:
+            raise Exception('Environment not set')
 
+        return self._environment
 
-@dataclass(frozen=True, order=True)
+    @environment.setter
+    def environment(self, env: Environment):
+        self._environment = env
+
+
+@dataclass(order=True)
 class SNum(SExp):
     """A lisp number"""
     value: int
 
     def __str__(self):
         return str(self.value)
 
+    def __hash__(self):
+        return hash(self.value)
 
-@dataclass(frozen=True)
+
+@dataclass
 class SSym(SExp):
     """A lisp symbol"""
     name: str
 
     def __str__(self):
         return self.name
 
+    def __hash__(self):
+        return hash(self.name)
+
 
 @dataclass
 class SVect(SExp):
-    """An n element vector"""
-    items: List[SExp]
+    """An n element vector
 
-    def to_list_items(self) -> Optional[List[SExp]]:
-        if len(self.items) == 0:
-            return []
-        if len(self.items) != 2:
-            return None
-        rest_items = self.items[1].to_list_items()
-        if rest_items is None:
-            return None
-        return [self.items[0]] + rest_items
+    >>> vect = SVect([
+    ...     SNum(42), SSym('spam'),
+    ...     SVect([SNum(43), SNum(44)])
+    ... ])
+    >>> str(vect)
+    '[42 spam [43 44]]'
+    >>>
+    >>> str(SVect([]))
+    '[]'
+    """
+    items: List[SExp]
 
     def __str__(self) -> str:
-        items = self.to_list_items()
-        if items is not None:
-            if len(items) == 2 and items[0] == SSym('quote'):
-                return f"'{items[1]}"
-            return f"({' '.join(str(i) for i in items)})"
         return f"[{' '.join(str(i) for i in self.items)}]"
 
 
-def to_slist(x: Iterable[SExp]) -> SExp:
-    acc = SVect([])
+@dataclass
+class SPair(SExp):
```

Comment:
Just checking: I presume you add this so that you can type pieces of syntax more explicitly as being pairs? So this here is for the syntax, and not as a value representation?

Reply:
Correct